src/types.rs

```diff
@@ -921,3 +921,87 @@ impl Debug for Log {
             .finish()
     }
 }
+
+/// An Ethereum Transaction Receipt
+#[derive(Clone, Eq, PartialEq, Deserialize, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct TransactionReceipt {
+    #[serde(rename = "type", flatten)]
+    pub kind: TransactionReceiptKind,
+
+    /// The hash of the transaction that emitted this log.
+    pub transaction_hash: Digest,
+    /// The index of the transaction that emitted this log within the block.
+    pub transaction_index: U256,
+    /// The hash of the block containing the transaction.
+    pub block_hash: Digest,
+    /// The height of the block containing the transaction.
+    pub block_number: U256,
+    /// Address of transaction sender.
+    pub from: Address,
+    /// Transaction receipient (Null for contract creation).
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub to: Option<Address>,
+    /// The sum of gas used by this transaction and all preceding transactions in the same block.
+    pub cumulative_gas_used: U256,
+    /// The amount of gas used for this specific transaction alone.
+    pub gas_used: U256,
+    /// Contract address created, or `None` if not a deployment.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub contract_address: Option<Address>,
+    /// Logs emitted by the transaction.
+    pub logs: Vec<Log>,
+    /// The log bloom filter.
+    pub logs_bloom: Bloom,
+    /// State root. Only present before activation of [EIP-658](https://eips.ethereum.org/EIPS/eip-658).
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub root: Option<Digest>,
+    /// Status: either 1 (success) or 0 (failure). Only present after activation of [EIP-658](https://eips.ethereum.org/EIPS/eip-658).
+    pub status: Option<U256>,
+}
+
+#[derive(Clone, Debug, Eq, PartialEq, Deserialize, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub enum TransactionReceiptKind {
+    #[serde(rename = "0x0")]
+    Legacy,
+    #[serde(rename = "0x1")]
+    Eip2930,
+    #[serde(rename = "0x2")]
+    Eip1559 {
+        /// The price paid post-execution by the transaction (i.e. base fee + priority fee).
+        /// Both fields in 1559-style transactions are *maximums* (max fee + max priority fee), the
+        /// amount that's actually paid by users can only be determined post-execution.
+        effective_gas_price: Option<U256>,
+    },
+    #[serde(rename = "0x3")]
+    Eip4844 {
+        /// The amount of blob gas used for this specific transaction.
+        /// Only specified for blob transactions as defined by EIP-4844.
+        blob_gas_used: U256,
+        /// The actual value per gas deducted from the sender's account for blob gas.
+        /// Only specified for blob transactions as defined by [EIP-4844](https://eips.ethereum.org/EIPS/eip-4844).
+        blob_gas_price: U256,
+    },
+}
```

Comment:
Were you able to confirm if this is actually needed? From my comment ☝️:

> **Do you know if post EIP-4844 if all transaction receipts will include the new field?** I wouldn't be surprised if it were the case (same thing happened post EIP-1559 with the effective_gas_price field being on all transaction receipts).
>
> **If not**, another potentially better strategy would be to have something like:
> ```
> enum TransactionReceiptKind {
>     Legacy,
>     Eip2930,
>     Eip1559,
>     Eip4844 {
>         blob_gas_price: U256,
>     }
> }
> ```

Reply:
I can't really confirm because I can't find any transactions of this type on even a testnet. I can try it out with another type that does exist.

Reply:
I was able to test that this approach (as is doesn't work as is) with transaction type 0x2 before moved the `effective_gas_price` field back into the main structure. Maybe I am missing something (tried to add `flatten` to serde tag)... but at this moment I can't really confirm anything. Not sure what to do here. Should we just exclude the Eip4844 stuff until it matters?

Reply:
> I can't really confirm because I can't find any transactions of this type on even a testnet.

I would read up on the EIP-4844 documentation and implementation (might be available in Geth and you might be able to enable it already in `--dev` mode to try it out).

Reply:
I'm seeing that is will only be there for the specific transaction type:

https://github.com/ethereum/go-ethereum/blob/4d9f3cd5d751efccd501b08ab6cf38a83b5e2858/core/types/receipt.go#L339-L343

This is logical since these types are transactions are meant for rollups and other transaction types will continue to be used.

Reply:
Or rather this: the fields are tagged with "omitempty" in the structure:

https://github.com/ethereum/go-ethereum/blob/4d9f3cd5d751efccd501b08ab6cf38a83b5e2858/core/types/receipt.go#L66C58-L67

Reply:
Cool, so they only show up for those transactions for now. Thanks for checking.